### PR TITLE
Collection expressions: verify element type when reporting conversion error to CollectionBuilder type

### PR DESCRIPTION
**Pull request by cston**
Fixes https://github.com/dotnet/roslyn/issues/69980

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs [hunk 1]
```diff
@@ -772,8 +772,14 @@ internal bool TryGetCollectionIterationType(ExpressionSyntax syntax, TypeSymbol
             if (collectionTypeKind == CollectionExpressionTypeKind.CollectionBuilder &&
                 elementType is null)
             {
-                Error(diagnostics, ErrorCode.ERR_CollectionBuilderNoElementType, node.Syntax, targetType);
-                return;
+                Debug.Assert(elementType is null); // GetCollectionExpressionTypeKind() does not set elementType for CollectionBuilder cases.
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs:775]: This was a condition of the `if`, why assert it immediately?

**Review by 333fred — APPROVED** [review]

File: src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs [hunk 2]
```diff
@@ -2851,6 +2851,138 @@ static void Main()
             comp.VerifyEmitDiagnostics();
         }
 
+        [Fact]
+        public void TypeInference_Spread_09()
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs:2855]: Are these new tests relevant to the implementation changes in this PR? It looks like the implementation change only affects error reporting.

**Review by RikkiGibson — APPROVED** [review]

File: src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs (hunk 2 above)
Comment (cston) [reply src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs:2855]: No, these tests are not related to the changes in the PR.
